File: UDPConnection.java
import java.io.IOException;
import java.net.*;
import java.nio.ByteBuffer;

public class UDPConnection {

    private DatagramSocket senderSocket;
    private DatagramSocket receiverSocket;
    private InetAddress targetAddress;

    public UDPConnection() {
        try {
            this.targetAddress = InetAddress.getByName("127.0.0.1");
            this.senderSocket = new DatagramSocket(null);
            this.receiverSocket = new DatagramSocket(null);
            this.senderSocket.setReuseAddress(true);
            this.receiverSocket.setReuseAddress(true);
            this.senderSocket = new DatagramSocket(7500);
            this.receiverSocket = new DatagramSocket(7501);
            this.receiverSocket.setSoTimeout(10);
        } catch (SocketException | UnknownHostException e) {
            System.err.println("Initialization failed: " + e.getMessage());
        }
    }

    public void sendTo(int value) {
        try {
            byte[] data = ByteBuffer.allocate(4).putInt(value).array(); //Length 4 bc 4 digit socket address
            DatagramPacket packet = new DatagramPacket(data, data.length, targetAddress, 7501);
            senderSocket.send(packet);
        } catch (IOException e) {
            System.err.println("Send error: " + e.getMessage());
        }
    }

    public Integer recvfrom() {
        byte[] buffer = new byte[4];
        DatagramPacket packet = new DatagramPacket(buffer, buffer.length);

        try {
            receiverSocket.receive(packet); // Will only wait for 10ms
            return ByteBuffer.wrap(packet.getData()).getInt();
        } catch (SocketTimeoutException e) {
            // No data was waiting; this is normal in a non-threaded loop
            return null; 
        } catch (IOException e) {
            System.err.println("Receive error: " + e.getMessage());
            return null;
        }
    }

    public void setNetworkAddress(String ipAddress) {
        try {
            this.targetAddress = InetAddress.getByName(ipAddress);
        } catch (UnknownHostException e) {
            System.err.println("Invalid IP.");
        }
    }
    
    // Test Main
    public static void main(String[] args) {
        UDPConnection comms = new UDPConnection();
        
        // Simulating a game loop
        for (int i = 0; i < 5; i++) {
            comms.sendTo(100 + i);
            
            // Manually check for data
            Integer input = comms.recvfrom();
            if (input != null) {
                System.out.println("Player equipment code: " + input);
            }
        }
    }
}

File: udp_connection.py
import socket
import struct


class UDPConnection:

    def __init__(self):
        try:
            self.target_address = "127.0.0.1"

            # Create sender socket (port 7500)
            self.sender_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            self.sender_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)

            # Create receiver socket (port 7501)
            self.receiver_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            self.receiver_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            self.receiver_socket.bind(("", 7501))

            # 10ms timeout (Java setSoTimeout(10))
            self.receiver_socket.settimeout(0.01)

        except Exception as e:
            print("Initialization failed:", e)

    def send_to(self, value):
        try:
            # Pack integer into 4 bytes 
            data = value.encode()
            self.sender_socket.sendto(data, (self.target_address, 7500))
        except Exception as e:
            print("Send error:", e)

    def recv_from(self):
        try:
            data, _ = self.receiver_socket.recvfrom(4096)
            return data.decode(errors='ignore')
        except socket.timeout:
            # Normal behavior (non-blocking style)
            return None
        except Exception as e:
            print("Receive error:", e)
            return None

    def set_network_address(self, ip_address):
        try:
            socket.inet_aton(ip_address)  # Validate IP
            self.target_address = ip_address
        except socket.error:
            print("Invalid IP.")